nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/pgp/PGPControllerService.java

```diff
@@ -0,0 +1,893 @@
+            final String keyFile = context.getProperty(PUBLIC_KEYRING_FILE).getValue();
+            try {
+                keys = readPublicKeys(new File(keyFile));
+            } catch (FileNotFoundException ignored) {
+            }
+        }
+
+        if (keys == null || keys.size() == 0)
+            return null;
+
+        if (context.getProperty(PUBLIC_KEY_USER_ID).isSet()) {
+            return keys.getPublicKey(context.getProperty(PUBLIC_KEY_USER_ID).getValue());
+        }
+
+        for (PGPPublicKey key : keys) {
+            if (key.isEncryptionKey())
+                return key;
+        }
+        return null;
+    }
+
+
+    private PGPPrivateKey getPrivateKey(PropertyContext context) {
+        String secretKeyUserId = null;
+        char[] privateKeyPassPhrase = null;
+        PGPSecretKeys keys = null;
+        PBESecretKeyDecryptor decryptor = null;
+
+        if (context.getProperty(SECRET_KEY_USER_ID).isSet()) {
+            secretKeyUserId = context.getProperty(SECRET_KEY_USER_ID).getValue();
+        }
+
+        if (context.getProperty(PRIVATE_KEY_PASS_PHRASE).isSet()) {
+            privateKeyPassPhrase = context.getProperty(PRIVATE_KEY_PASS_PHRASE).getValue().toCharArray();
+
+            try {
+                decryptor = new JcePBESecretKeyDecryptorBuilder().setProvider("BC").build(privateKeyPassPhrase);
+            } catch (PGPException e) {
+                return null;
+            }
+        }
+
+        if (context.getProperty(SECRET_KEYRING_TEXT).isSet()) {
+            final String secretKeyText = context.getProperty(SECRET_KEYRING_TEXT).getValue();
+            keys = readSecretKeys(secretKeyText);
+
+        } else  if (context.getProperty(SECRET_KEYRING_FILE).isSet()) {
+            try {
+                keys = readSecretKeys(new File(context.getProperty(SECRET_KEYRING_FILE).getValue()));
+            } catch (FileNotFoundException e) {
+                return null;
+            }
+        }
+
+        if (keys == null || keys.size() == 0)
+            return null;
+
+        if (secretKeyUserId != null) {
+            try {
+                return keys.getSecretKey(secretKeyUserId).extractPrivateKey(decryptor);
+            } catch (PGPException e) {
+                return null;
+            }
+        }
+
+        for (PGPSecretKey key : keys) {
+            if (!key.isPrivateKeyEmpty()) {
+                try {
+                    return key.extractPrivateKey(decryptor);
+                } catch (final Exception ignored) {
+                    // pass
+                }
+            }
+        }
+        return null; // no private key available from the given config
+    }
+
+
+    private char[] getPBEPassPhrase(PropertyContext context) {
+        if (context.getProperty(PBE_PASS_PHRASE).isSet()) {
+            return context.getProperty(PBE_PASS_PHRASE).getValue().toCharArray();
+        }
+        return null;
+    }
+
+
+    static class PGPPublicKeys extends ArrayList<PGPPublicKey> {
```

Comment:
Is this used anywhere a `List<>` is expected? If we used `HashMap<Long, PGPPublicKey>` or `HashMap<String, PGPPublicKey>` this would provide `O(1)` for lookup (for whichever ID type we use as the key).

Reply:
Migrated these types to mapping types.  On read, we're now using all of the user id values as keys (the user id field is a composite type).  Lookup happens as expected when specifying a key by user id via `get`, and the existing lookup methods (by string) now defer to that.